// @forgerock/sdk-types/dist/src/lib/legacy-mware.types.js
/*
 * Copyright (c) 2025 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
/** *************************************************************
 * Types for legacy middleware configuration
 * **************************************************************/
export var LegacyActionTypes;
(function (LegacyActionTypes) {
    LegacyActionTypes["Authenticate"] = "AUTHENTICATE";
    LegacyActionTypes["Authorize"] = "AUTHORIZE";
    LegacyActionTypes["EndSession"] = "END_SESSION";
    LegacyActionTypes["Logout"] = "LOGOUT";
    LegacyActionTypes["ExchangeToken"] = "EXCHANGE_TOKEN";
    LegacyActionTypes["RefreshToken"] = "REFRESH_TOKEN";
    LegacyActionTypes["ResumeAuthenticate"] = "RESUME_AUTHENTICATE";
    LegacyActionTypes["RevokeToken"] = "REVOKE_TOKEN";
    LegacyActionTypes["StartAuthenticate"] = "START_AUTHENTICATE";
    LegacyActionTypes["UserInfo"] = "USER_INFO";
    LegacyActionTypes["WellKnown"] = "WELL_KNOWN";
})(LegacyActionTypes || (LegacyActionTypes = {}));
