// @forgerock/davinci-client/dist/src/lib/client.store.utils.js
/*
 * Copyright (c) 2025 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
import { configureStore } from '@reduxjs/toolkit';
import { configSlice } from './config.slice.js';
import { nodeSlice } from './node.slice.js';
import { davinciApi } from './davinci.api.js';
import { wellknownApi } from './wellknown.api.js';
export function createClientStore({ requestMiddleware, logger, }) {
    return configureStore({
        reducer: {
            config: configSlice.reducer,
            node: nodeSlice.reducer,
            [davinciApi.reducerPath]: davinciApi.reducer,
            [wellknownApi.reducerPath]: wellknownApi.reducer,
        },
        middleware: (getDefaultMiddleware) => getDefaultMiddleware({
            thunk: {
                extraArgument: {
                    /**
                     * This becomes the `api.extra` argument, and will be passed into the
                     * customer query wrapper for `baseQuery`
                     */
                    requestMiddleware,
                    logger,
                },
            },
        })
            .concat(davinciApi.middleware)
            .concat(wellknownApi.middleware),
    });
}
export function handleUpdateValidateError(message, type, cb) {
    cb(message);
    return function () {
        return {
            error: {
                message: message,
                type: type,
            },
            type: 'internal_error',
        };
    };
}
