// @forgerock/davinci-client/dist/src/lib/fido/fido.js
/*
 * Copyright (c) 2025 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
import { Micro } from 'effect';
import { exitIsFail, exitIsSuccess } from 'effect/Micro';
import { transformAssertion, transformAuthenticationOptions, transformPublicKeyCredential, transformRegistrationOptions, } from './fido.utils.js';
/**
 * A client function that returns a set of methods for transforming DaVinci data and
 * interacting with the WebAuthn API for registration and authentication
 * @function fido
 * @returns {FidoClient} - A set of methods for FIDO registration and authentication
 */
export function fido() {
    return {
        /**
         * Call WebAuthn API to create keypair and get public key credential
         */
        register: async function register(options) {
            const createCredentialµ = Micro.sync(() => transformRegistrationOptions(options)).pipe(Micro.flatMap((publicKeyCredentialCreationOptions) => Micro.tryPromise({
                try: () => navigator.credentials.create({
                    publicKey: publicKeyCredentialCreationOptions,
                }),
                catch: (error) => {
                    console.error('Failed to create keypair: ', error);
                    return {
                        error: 'registration_error',
                        message: 'FIDO registration failed',
                        type: 'fido_error',
                    };
                },
            })), Micro.flatMap((credential) => {
                if (!credential) {
                    return Micro.fail({
                        error: 'registration_error',
                        message: 'FIDO registration failed: No credential returned',
                        type: 'fido_error',
                    });
                }
                else {
                    const formattedCredential = transformPublicKeyCredential(credential);
                    return Micro.succeed(formattedCredential);
                }
            }));
            const result = await Micro.runPromiseExit(createCredentialµ);
            if (exitIsSuccess(result)) {
                return result.value;
            }
            else if (exitIsFail(result)) {
                return result.cause.error;
            }
            else {
                return {
                    error: 'fido_registration_error',
                    message: result.cause.message,
                    type: 'unknown_error',
                };
            }
        },
        /**
         * Call WebAuthn API to get assertion
         */
        authenticate: async function authenticate(options) {
            const getAssertionµ = Micro.sync(() => transformAuthenticationOptions(options)).pipe(Micro.flatMap((publicKeyCredentialRequestOptions) => Micro.tryPromise({
                try: () => navigator.credentials.get({
                    publicKey: publicKeyCredentialRequestOptions,
                }),
                catch: (error) => {
                    console.error('Failed to authenticate: ', error);
                    return {
                        error: 'authentication_error',
                        message: 'FIDO authentication failed',
                        type: 'fido_error',
                    };
                },
            })), Micro.flatMap((assertion) => {
                if (!assertion) {
                    return Micro.fail({
                        error: 'authentication_error',
                        message: 'FIDO authentication failed: No credential returned',
                        type: 'fido_error',
                    });
                }
                else {
                    const formattedAssertion = transformAssertion(assertion);
                    return Micro.succeed(formattedAssertion);
                }
            }));
            const result = await Micro.runPromiseExit(getAssertionµ);
            if (exitIsSuccess(result)) {
                return result.value;
            }
            else if (exitIsFail(result)) {
                return result.cause.error;
            }
            else {
                return {
                    error: 'fido_authentication_error',
                    message: result.cause.message,
                    type: 'unknown_error',
                };
            }
        },
    };
}
